contracts/Staking.sol

```diff
@@ -0,0 +1,180 @@
+        address _SEURO_ADDRESS,
+        uint256 _TOTAL_SEURO,
+        uint _SEUROTST,
+        uint _INTEREST
+    ) external onlyOwner {
+
+        // CHORE needs refactor
+        require(active == false, 'err-already-active');
+        require(initialised == 0, 'err-already-initialised');
+        require(_end > _start, 'err-start-end');
+        require(_end >= block.timestamp, 'err-invalid-end');
+
+        TOTAL_SEURO = _TOTAL_SEURO;
+        SEURO_REMAINING = _TOTAL_SEURO;
+        SEUROTST = _SEUROTST;
+        INTEREST = _INTEREST;
+        TST_ADDRESS = _TST_ADDRESS;
+        SEURO_ADDRESS = _SEURO_ADDRESS;
+        startTime = _start;
+        endTime = _end;
+        initialised = block.timestamp;
+
+        // TODO variable
+        minTST = 1 ether;
+
+        active = true;
+    }
+
+    function disable() external onlyOwner {
+        require(active, 'err-not-active');
+        active = false;
+    }
+
+    // reward works out the amount of seuro given to the user based on the
+    // amount of TST they first put in.
+    function reward(uint256 _amount) public view returns (uint256) {
+        uint256 SEURO = _amount * SEUROTST / 10_000;
+        uint256 REWARD = SEURO * INTEREST / 10_000;
+        return REWARD + SEURO;
+    }
+
+    function mint(uint256 _amount) external returns(uint256) {
+
+        // TODO CHORE needs refactor
+        require(active == true, 'err-not-active');
+        require(_amount >= minTST, 'err-not-min');
+        require(block.timestamp >= startTime, 'err-not-started');
+        require(block.timestamp < endTime, 'err-finished');
+
+        // calculate the reward so we can also update the remaining SEURO
+        uint256 total = reward(_amount);
+        require(SEURO_REMAINING >= total, 'err-overlimit');
+
+        // Transfer funds from sender to this contract
+        // TODO send to some other guy not this contract!
+
+        IERC20 TOKEN = IERC20(TST_ADDRESS);
+        TOKEN.transferFrom(msg.sender, address(this), _amount);
+
+        // fetch current tokenID
+        uint256 newItemId = _tokenIds.current();
+
+        Position memory position = _positions[msg.sender];
+
+        if (position.nonce > 0) {
+            position.totalValue += _amount;
+            position.nonce += 1;
+            position.reward += total;
+        }
+
+        if (position.nonce == 0) {
+            _mint(msg.sender, newItemId);
+
+            position.nonce = 1;
+            position.open = true;
+            position.totalValue = _amount;
+            position.tokenId = newItemId;
+            position.reward = total;
+
+            _tokenIds.increment();
+        }
+
+        // update the position
+        _positions[msg.sender] = position;
+
+        // update the remaining SEURO
+        SEURO_REMAINING -= total;
+
+        // makes no sence to return this.
+        return newItemId;
+    }
+
+    function burn() public {
+        require(block.timestamp >= endTime, 'err-pool-open');
```

Comment:
we only allow users to get their reward after the endTime yes, and yes it would be the same for each user. 

So, if a user tries to their reward before the endTime, they would get nada.

Reply:
Can you review the flow chart again to make sure you know what I'm implementing

Reply:
removed